lib/shared/sdk/usbboot/index.js

```diff
@@ -193,251 +133,318 @@ const usbIdToString = (id) => {
-        debug('Couldn\'t claim the interface. Assuming the device is gone')
-        return null
+  /**
+   * @summary Write bootcode to USB device (usbboot first stage)
+   * @private
+   *
+   * @description
+   * After this stage is run, the USB will be re-mounted as 0x0a5c:0x2764.
+   *
+   * @param {Object} device - node-usb device
+   * @param {Object} endpoint - node-usb endpoint
+   * @param {Buffer} bootCodeBuffer - bootcode buffer
+   * @returns {Promise}
+   *
+   * @example
+   * const usb = require('usb')
+   * const device = usb.findByIds(0x0a5c, 0x2763)
+   * const bootcode = fs.readFileSync('./bootcode.bin')
+   *
+   * adapter.writeBootCode(device, device.interfaces(0).endpoint(1), bootcode).then(() => {
+   *   console.log('Done!')
+   * })
+   */
+  static writeBootCode (device, endpoint, bootCodeBuffer) {
+    debug('Writing bootcode')
+    debug(`Bootcode buffer length: ${bootCodeBuffer.length}`)
+    const bootMessageBuffer = protocol.createBootMessageBuffer(bootCodeBuffer.length)
+
+    debug('Writing boot message buffer to out endpoint')
+    return protocol.write(device, endpoint, bootMessageBuffer).then(() => {
+      debug('Writing boot code buffer to out endpoint')
+      return protocol.write(device, endpoint, bootCodeBuffer)
+    }).then(() => {
+      debug('Reading return code from device')
+      return protocol.read(device, protocol.RETURN_CODE_LENGTH)
+    }).then((data) => {
+      const returnCode = data.readInt32LE()
+      debug(`Received return code: ${returnCode}`)
+
+      if (returnCode !== protocol.RETURN_CODE_SUCCESS) {
+        throw new Error(`Couldn't write the bootcode, got return code ${returnCode} from device`)
       }
+    })
+  }
 
-      throw error
-    }
-
-    const endpoint = deviceInterface.endpoint(addresses.endpoint)
-
-    return Bluebird.try(() => {
-      const serialNumberIndex = device.deviceDescriptor.iSerialNumber
-      debug(`Serial number index: ${serialNumberIndex}`)
+  /**
+   * @summary Mount a USB device as a block device (usbboot second stage)
+   * @private
+   *
+   * @description
+   * The possible files you can pass here are:
+   *
+   * - autoboot.txt
+   * - config.txt
+   * - recovery.elf
+   * - start.elf
+   * - fixup.dat
+   *
+   * @param {Object} device - node-usb device
+   * @param {Object} endpoint - node-usb endpoint
+   * @param {Function} readFile - read file function
+   * @returns {Promise}
+   *
+   * @example
+   * const fs = Bluebird.promisifyAll(require('fs'))
+   * const usb = require('usb')
+   * const device = usb.findByIds(0x0a5c, 0x2763)
+   *
+   * adapter.startFileServer(device, device.interfaces(0).endpoint(1), (name) => {
+   *   return fs.readFileAsync(name)
+   * }).then(() => {
+   *   console.log('Done!')
+   * })
+   */
+  startFileServer (device, endpoint, readFile) {
+    debug('Listening for file messages')
+    return protocol
+        .read(device, protocol.FILE_MESSAGE_SIZE)
```

Comment:
I think there is way too much indentation here.

Reply:
Oh yes, fixed now!